File: documentapi/src/vespa/documentapi/messagebus/policies/localservicepolicy.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "localservicepolicy.h"
#include <vespa/documentapi/messagebus/documentprotocol.h>
#include <vespa/messagebus/routing/verbatimdirective.h>
#include <vespa/messagebus/messagebus.h>
#include <vespa/vespalib/util/stringfmt.h>
#include <vespa/log/log.h>
LOG_SETUP(".localservicepolicy");

namespace documentapi {

LocalServicePolicy::CacheEntry::CacheEntry() :
    _offset(0),
    _generation(0),
    _recipients()
{
}

LocalServicePolicy::LocalServicePolicy(const string &param) :
    _lock(),
    _address(param),
    _cache()
{
}

LocalServicePolicy::~LocalServicePolicy() = default;

void
LocalServicePolicy::select(mbus::RoutingContext &ctx)
{
    mbus::Route route = ctx.getRoute();
    route.setHop(0, getRecipient(ctx));
    ctx.addChild(route);
}

void
LocalServicePolicy::merge(mbus::RoutingContext &context)
{
    DocumentProtocol::merge(context);
}

string
LocalServicePolicy::getCacheKey(const mbus::RoutingContext &ctx) const
{
    return ctx.getRoute().getHop(0).toString();
}

mbus::Hop
LocalServicePolicy::getRecipient(mbus::RoutingContext &ctx)
{
    vespalib::LockGuard guard(_lock);
    CacheEntry &entry = update(ctx);
    if (entry._recipients.empty()) {
        mbus::Hop hop = ctx.getRoute().getHop(0);
        hop.setDirective(ctx.getDirectiveIndex(), std::make_shared<mbus::VerbatimDirective>("*"));
        return hop;
    }
    if (++entry._offset >= entry._recipients.size()) {
        entry._offset = 0;
    }
    return entry._recipients[entry._offset];
}

LocalServicePolicy::CacheEntry &
LocalServicePolicy::update(mbus::RoutingContext &ctx)
{
    uint32_t upd = ctx.getMirror().updates();
    CacheEntry &entry = _cache.insert(std::map<string, CacheEntry>::value_type(getCacheKey(ctx), CacheEntry())).first->second;
    if (entry._generation != upd) {
        entry._generation = upd;
        entry._recipients.clear();

        string pattern = vespalib::make_string("%s*%s",
                                                    ctx.getHopPrefix().c_str(),
                                                    ctx.getHopSuffix().c_str());
        slobrok::api::IMirrorAPI::SpecList entries = ctx.getMirror().lookup(pattern);

        string self = _address.empty() ? toAddress(ctx.getMessageBus().getConnectionSpec()) : _address;
        for (slobrok::api::IMirrorAPI::SpecList::iterator it = entries.begin();
             it != entries.end(); ++it)
        {
            LOG(debug, "Matching self '%s' to '%s'.", self.c_str(), it->second.c_str());
            if (self == toAddress(it->second)) {
                LOG(debug, "Match, add it");
                entry._recipients.push_back(mbus::Hop::parse(it->first));
            }
        }
    }
    return entry;
}

string
LocalServicePolicy::toAddress(const string &connection)
{
    if (connection.substr(0, 4) == "tcp/") {
        uint32_t pos = connection.find_first_of(':', 4);
        if (pos > 4) {
            return connection.substr(4, pos - 4);
        }
    }
    return "";
}

}

File: documentapi/src/main/java/com/yahoo/documentapi/messagebus/protocol/LocalServicePolicy.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.documentapi.messagebus.protocol;

import com.yahoo.jrt.slobrok.api.Mirror;
import com.yahoo.messagebus.routing.Hop;
import com.yahoo.messagebus.routing.Route;
import com.yahoo.messagebus.routing.RoutingContext;
import com.yahoo.messagebus.routing.VerbatimDirective;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * This policy implements the logic to prefer local services that matches a slobrok pattern.
 *
 * @author Simon Thoresen Hult
 */
public class LocalServicePolicy implements DocumentProtocolRoutingPolicy {

    private final String localAddress;
    private final Map<String, CacheEntry> cache = new HashMap<>();

    /**
     * Constructs a policy that will choose local services that match the slobrok pattern in which this policy occured.
     * If no local service can be found, this policy simply returns the asterisk to allow the network to choose any.
     *
     * @param param The address to use for this, if empty this will resolve to hostname.
     */
    LocalServicePolicy(String param) {
        localAddress = (param != null && param.length() > 0) ? param : null;
    }

    public void select(RoutingContext ctx) {
        Route route = new Route(ctx.getRoute());
        route.setHop(0, getRecipient(ctx));
        ctx.addChild(route);
    }

    public void merge(RoutingContext ctx) {
        DocumentProtocol.merge(ctx);
    }

    /**
     * Returns the appropriate recipient hop for the given routing context. This method provides synchronized access to
     * the internal cache.
     *
     * @param ctx The routing context.
     * @return The recipient hop to use.
     */
    private synchronized Hop getRecipient(RoutingContext ctx) {
        CacheEntry entry = update(ctx);
        if (entry.recipients.isEmpty()) {
            Hop hop = new Hop(ctx.getRoute().getHop(0));
            hop.setDirective(ctx.getDirectiveIndex(), new VerbatimDirective("*"));
            return hop;
        }
        if (++entry.offset >= entry.recipients.size()) {
            entry.offset = 0;
        }
        return new Hop(entry.recipients.get(entry.offset));
    }

    /**
     * Updates and returns the cache entry for the given routing context. This method assumes that synchronization is
     * handled outside of it.
     *
     * @param ctx The routing context.
     * @return The updated cache entry.
     */
    private CacheEntry update(RoutingContext ctx) {
        String key = getCacheKey(ctx);
        CacheEntry entry = cache.get(key);
        if (entry == null) {
            entry = new CacheEntry();
            cache.put(key, entry);
        }
        int upd = ctx.getMirror().updates();
        if (entry.generation != upd) {
            entry.generation = upd;
            entry.recipients.clear();

            List<Mirror.Entry> arr = ctx.getMirror().lookup(ctx.getHopPrefix() + "*" + ctx.getHopSuffix());
            String self = localAddress != null ? localAddress : toAddress(ctx.getMessageBus().getConnectionSpec());
            for (Mirror.Entry item : arr) {
                if (self.equals(toAddress(item.getSpecString()))) {
                    entry.recipients.add(Hop.parse(item.getName()));
                }
            }
        }
        return entry;
    }

    /**
     * Returns a cache key for this instance of the policy. Because behaviour is based on the hop in which the policy
     * occurs, the cache key is the hop string itself.
     *
     * @param ctx The routing context.
     * @return The cache key.
     */
    private String getCacheKey(RoutingContext ctx) {
        return ctx.getRoute().getHop(0).toString();
    }

    /**
     * Defines the necessary cache data.
     */
    private class CacheEntry {
        private final List<Hop> recipients = new ArrayList<>();
        private int generation = 0;
        private int offset = 0;
    }

    /**
     * Searches the given connection spec for a hostname or IP address. If an address is not found, this method returns
     * null.
     *
     * @param connection The connection spec to search.
     * @return The address, may be null.
     */
    private static String toAddress(String connection) {
        if (connection.startsWith("tcp/")) {
            int pos = connection.indexOf(':');
            if (pos > 4) {
                return connection.substring(4, pos);
            }
        }
        return null;
    }

    public void destroy() {
    }
}
